### ISSUE TITLE
Feature Request: Group By function?

### ISSUE THREAD

**Issue by StevenACoffman** [bradenaw/juniper#4]
It would be nice to have a Group-By function for collections like https://www.scala-lang.org/api/2.12.7/scala/collection/Iterable.html#groupBy[K](f:A=%3EK):scala.collection.immutable.Map[K,Repr]
or (in Go before generics):
https://github.com/ahmetb/go-linq/blob/master/groupby.go

**Comment by bradenaw** [comment]
Is there a method signature in Go you're imagining?

There's an xslices.Group that's nearly similar, although it sounds like you'd perhaps prefer iterating each group in parallel without allocating. Is that right?

**Comment by StevenACoffman** [comment]
That is what I was thinking, but now I'm questioning how it would be possible to avoid intermediate allocations for the individual groups. 🤔

**Comment by bradenaw** [comment]
In the general case, I don't think so.

`iterator.Runs` will do it with no allocations if the groups are contiguous.

**Comment by StevenACoffman** [comment]
Ok, I'm closing this issue, but if someone else comes up with a brilliant idea, feel free to open a new one! Thanks for sharing your work!
